streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/TaskMovement.java

```diff
@@ -16,128 +16,103 @@
-                    log.error("Task {} is assigned to client {} in initial assignment but has no owner in the final " +
-                                  "balanced assignment.", task, source);
-                    throw new IllegalStateException("Found task in initial assignment that was not assigned in the final.");
-                } else if (!source.equals(destination)) {
-                    if (destinationClientIsCaughtUp(task, destination, tasksToCaughtUpClients)) {
-                        sourceClientTasksIterator.remove();
-                        statefulActiveTaskAssignment.get(destination).add(task);
-                    } else {
-                        if (clientStates.get(destination).prevStandbyTasks().contains(task)
-                                && tasksToRemainingStandbys.get(task) > 0
-                        ) {
-                            decrementRemainingStandbys(task, tasksToRemainingStandbys);
-                        } else {
-                            --remainingAllowedWarmupReplicas;
-                        }
+            final ClientState sourceClientState = clientStates.get(sourceClient);
+            sourceClientState.assignActive(movement.task);
+            clientsByTaskLoad.offer(sourceClient);
 
-                        movements.add(new TaskMovement(task, source, destination));
-                        if (remainingAllowedWarmupReplicas == 0) {
-                            return movements;
-                        }
-                    }
-                }
+            final ClientState destinationClientState = clientStates.get(movement.destination);
+            if (shouldAssignWarmupReplica(movement.task, destinationClientState, remainingWarmupReplicas, tasksToRemainingStandbys)) {
+                destinationClientState.assignStandby(movement.task);
+                clientsByTaskLoad.offer(movement.destination);
+                warmupReplicasAssigned = true;
             }
         }
-        return movements;
+        return warmupReplicasAssigned;
     }
 
-    private static boolean destinationClientIsCaughtUp(final TaskId task,
-                                                       final UUID destination,
-                                                       final Map<TaskId, SortedSet<UUID>> tasksToCaughtUpClients) {
-        final Set<UUID> caughtUpClients = tasksToCaughtUpClients.get(task);
-        return caughtUpClients != null && caughtUpClients.contains(destination);
+    private static boolean shouldAssignWarmupReplica(final TaskId task,
+                                                     final ClientState destinationClientState,
+                                                     final AtomicInteger remainingWarmupReplicas,
+                                                     final Map<TaskId, Integer> tasksToRemainingStandbys) {
+        if (destinationClientState.previousAssignedTasks().contains(task) && tasksToRemainingStandbys.get(task) > 0) {
```

Comment:
This question from my previous review went unnoticed (or you did simply not care ;-)).

>  Q: Why do we even care at all whether the task was running on the client? What if we just assign a real stand-by task if we have a spare one?

Reply:
I think I answered this already. We're trying not to decrease the overall availability the standbys are providing, which could happen if we drop a caught-up standby in order to warm up an empty node. We can certainly do better than what we do now, which is not very efficient in terms of task movement, but I think it's good enough for this PR.